go/ephemeral/ephemeral.go

```diff
@@ -188,13 +193,236 @@ func GetOwnDeviceEKs(ctx context.Context, g *libkb.GlobalContext) ([]keybase1.De
+	if err != nil {
+		return data, err
+	}
+
+	// TODO: Store the key.
+
+	dhKeypair, err := seed.DeriveDHKey()
+	if err != nil {
+		return data, err
+	}
+	metadata := keybase1.UserEkMetadata{
+		Kid:        dhKeypair.GetKID(),
+		Generation: generation,
+		HashMeta:   currentMerkleRoot.HashMeta(),
+	}
+	metadataJSON, err := json.Marshal(metadata)
+	if err != nil {
+		return data, err
+	}
+
+	// Sign the metadata blob with the latest PUK.
+	pukKeyring, err := g.GetPerUserKeyring()
+	if err != nil {
+		return data, err
+	}
+	signingKey, err := pukKeyring.GetLatestSigningKey(ctx)
+	if err != nil {
+		return data, err
+	}
+	signedPacket, _, err := signingKey.SignToString(metadataJSON)
+
+	// Box the seed up for each valid deviceEK.
+	ownDeviceEKs, err := GetOwnDeviceEKs(ctx, g)
+	if err != nil {
+		return data, err
+	}
+	boxes := []UserEKBoxMetadata{}
+	for deviceID, deviceEK := range ownDeviceEKs {
+		recipientKey, err := libkb.ImportKeypairFromKID(deviceEK.Kid)
+		if err != nil {
+			return data, err
+		}
+		// Encrypting with a nil sender means we'll generate a random sender private key.
+		box, err := recipientKey.EncryptToString(seed[:], nil)
+		if err != nil {
+			return data, err
+		}
+		boxMetadata := UserEKBoxMetadata{
+			RecipientDeviceID:   deviceID,
+			RecipientGeneration: deviceEK.Generation,
+			Box:                 box,
+		}
+		boxes = append(boxes, boxMetadata)
+	}
+
+	err = postNewUserEK(ctx, g, signedPacket, boxes)
+	if err != nil {
+		return data, err
+	}
+
+	return metadata, nil
+}
+
+type UserEKResponse struct {
+	Result struct {
+		MerklePayload string `json:"merkle_payload"`
+		Sig           string `json:"sig"`
+	} `json:"result"`
+}
+
+// Returns nil if there is no active userEK, either because none has ever been
+// created or because the user has gone stale.
+func GetActiveUserEK(ctx context.Context, g *libkb.GlobalContext, uid keybase1.UID) (*keybase1.UserEkMetadata, error) {
+	apiArg := libkb.APIArg{
+		Endpoint:    "user/user_ek",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		NetContext:  ctx,
+		Args:        libkb.HTTPArgs{},
+	}
+	res, err := g.GetAPI().Get(apiArg)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedResponse := UserEKResponse{}
+	err = res.Body.UnmarshalAgain(&parsedResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	// The client now needs to verify several things about the blob its
+	// received:
+	// 1) Each is validly signed.
+	// 2) The signing key is the user's latest PUK.
+	// 3) The merkle payload supplied by the server matches the hash that's
+	//    been signed over.
+	// 4) The key hasn't expired. That is, the Merkle root it was delegated
+	//    with is within one week of the current root. The server deliberately
+	//    avoids doing this filtering for us, and finding expired keys in the
+	//    results here is expected. We silently drop them.
+
+	// Verify the sig.
+	signerKey, payload, _, err := libkb.NaclVerifyAndExtract(parsedResponse.Result.Sig)
```

Comment:
let's also move this to a helper since when we will want to validate the sig when we get the box as well

Reply:
Something like:
`func verifyUserEKSig(sig string) (verified bool, payload keybase1.UserEkMetadata, err error)`

or we can drop the bool and just return an error if the verification fails